auth/ldap/ldap.go

```diff
@@ -65,16 +65,65 @@ func (l *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
 		return nil, errors.New("Can not get any available LDAP_BASE_DN.")
 	}
 
-	baseDn := fmt.Sprintf(ldapBaseDn, m.Principal)
-	log.Debug("baseDn:", baseDn)
+	//baseDn := fmt.Sprintf(ldapBaseDn, m.Principal)
+	log.Debug("baseDn:", ldapBaseDn)
+	
+	filter := os.Getenv("LDAP_FILTER")
+	log.Debug("ldapfilter:", filter)
+	uid := os.Getenv("LDAP_UID")
+	if filter != "" {
+		filter = "(&("+filter + ")("+ uid + "=" + m.Principal + "))"
+	}else {
+		filter ="(" + uid + "=" + m.Principal +")"
+	}
+	log.Debug("2 filter", filter)
+	ldapScope := os.Getenv("LDAP_SCOPE")
+	var scope int
+	if ldapScope == "1" {
+		scope = openldap.LDAP_SCOPE_BASE
+	}else if ldapScope == "2" {
+		scope = openldap.LDAP_SCOPE_ONELEVEL
+	}else {
+		scope = openldap.LDAP_SCOPE_SUBTREE
+	}
+	attributes := []string{"cn", "mail", "email", "uid"}
+	result, err := ldap.SearchAll(ldapBaseDn, scope, filter, attributes)
+	if err != nil {
+		return nil, err
+	}
+	if len(result.Entries()) != 1 {
+		log.Warningf("Found more than one entry.")
+		return nil, nil
+	}
+	en := result.Entries()[0]
+	bindDN := en.Dn()
+	u := models.User{}
+	for _, attr := range en.Attributes() {
```

Comment:
So is it possible the entry has no uid nor cn? we probably should check that and fail before inserting to DB?  or shall we use the uid attribute user set in the .cfg as username?